source/kernel/src/services/sdmmc.rs

```diff
@@ -0,0 +1,489 @@
+            .and_then(|resp| resp.body)
+    }
+
+    /// Reset the card
+    pub async fn reset(&mut self) -> Result<(), Error> {
+        self.cmd(Command::default()).await.map(|_| ())
+    }
+
+    /// Initialize the card
+    pub async fn initialize(&mut self) -> Result<CardType, Error> {
+        /// Request switch to 1.8V
+        #[allow(dead_code)]
+        const OCR_S18R: u32 = 0x1000000;
+        /// Host supports high capacity
+        const OCR_HCS: u32 = 0x40000000;
+        /// Card has finished power up routine if bit is high
+        const OCR_NBUSY: u32 = 0x80000000;
+        /// Valid bits for voltage setting
+        const OCR_VOLTAGE_MASK: u32 = 0x007FFF80;
+
+        let mut card_type = CardType::SD1;
+
+        match self
+            .cmd(Command {
+                index: 8,
+                argument: 0x1AA,
+                options: HardwareOptions::None,
+                kind: CommandKind::Control,
+                rsp_type: ResponseType::Short,
+                rsp_crc: true,
+                buffer: None,
+            })
+            .await?
+        {
+            Response::Short { value, .. } => {
+                tracing::trace!("CMD8 response: {value:#x}");
+                if value == 0x1AA {
+                    card_type = CardType::SD2;
+                }
+            }
+            Response::Long(_) => return Err(Error::Response),
+        }
+
+        // TODO: limit the number of attempts
+        let ocr = loop {
+            // Go to *APP* mode before sending application command
+            let _ = self
+                .cmd(Command {
+                    index: 55,
+                    argument: 0,
+                    options: HardwareOptions::None,
+                    kind: CommandKind::Control,
+                    rsp_type: ResponseType::Short,
+                    rsp_crc: true,
+                    buffer: None,
+                })
+                .await?;
+
+            let mut op_cond_arg = OCR_VOLTAGE_MASK & 0x00ff8000;
+            if card_type != CardType::SD1 {
+                op_cond_arg |= OCR_HCS;
+            }
+            match self
+                .cmd(Command {
+                    index: 41,
+                    argument: op_cond_arg,
+                    options: HardwareOptions::None,
+                    kind: CommandKind::Control,
+                    rsp_type: ResponseType::Short,
+                    rsp_crc: false,
+                    buffer: None,
+                })
+                .await?
+            {
+                Response::Short { value, .. } => {
+                    tracing::trace!("ACMD41 response: {value:#x}");
+                    if value & OCR_NBUSY == OCR_NBUSY {
+                        // Card has finished power up, data is valid
+                        break value;
+                    }
+                }
+                Response::Long(_) => return Err(Error::Response),
+            }
+
+            // TODO: wait some time (e.g., 1ms?)
```

Comment:
we do have a globally available kernel timer...

Reply:
Is this using `maitake::time::sleep(...).await`?